### PR TITLE
kv: complete the implementation of the new memdb

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve?
This is the second patchset of #18650. Please see the full description in the original PR page.

Contains `arena`, `allocator`, `vlog` and `iterator`.

### Release note 

-  No release note

### CODE REVIEW

File: kv/new_memdb_arena.go [hunk 1]
```diff
@@ -35,11 +36,13 @@ func (addr memdbArenaAddr) isNull() bool {
...
+	endian.PutUint32(dst[4:], addr.off)
 }
 
 func (addr *memdbArenaAddr) load(src []byte) {
-	panic("todo")
+	addr.idx = endian.Uint32(src)
+	addr.off = endian.Uint32(src[4:])
 }
 
 type memdbArena struct {
```
Comment (bobotu) [inline kv/new_memdb_arena.go:48]: `memdbArena` is copied from the existing `Sandbox`'s arena.

File: kv/new_memdb_arena.go [hunk 2]
```diff
@@ -104,25 +166,52 @@ type nodeAllocator struct {
...
+	n.klen = uint16(len(key))
+	copy(n.getKey(), key)
+	return addr, n
 }
 
-func (a *nodeAllocator) freeNode(addr memdbArenaAddr) {}
+var testMode = false
+
+func (a *nodeAllocator) freeNode(addr memdbArenaAddr) {
+	if testMode {
```
Comment (coocood) [inline kv/new_memdb_arena.go:198]: Can we make it a field of the allocator to avoid race?

Comment (bobotu) [reply kv/new_memdb_arena.go:198]: It will be set in test package's `init`. No race, and be enabled for all tests in other packages.

**Comment by bobotu** [comment]
PTAL @coocood @tiancaiamao

File: kv/new_memdb_arena.go [hunk 3]
```diff
@@ -137,22 +226,118 @@ type memdbVlogHdr struct {
...
 
 func (l *memdbVlog) revertToCheckpoint(db *memdb, cp *memdbCheckpoint) {
-	panic("todo")
+	cursor := l.checkpoint()
+	for !cp.isSamePosition(&cursor) {
+		hdrOff := cursor.offsetInBlock - memdbVlogHdrSize
+		block := l.blocks[cursor.blocks-1].buf
+		var hdr memdbVlogHdr
+		hdr.load(block[hdrOff:])
+		hdr.load(block[hdrOff:])
```
Comment (coocood) [inline kv/new_memdb_arena.go:277]: duplicated code?

File: kv/new_memdb_arena.go [hunk 4]
```diff
@@ -137,22 +226,117 @@ type memdbVlogHdr struct {
...
+				db.deleteNode(node)
+			} else {
+				node.setKeyFlags(keptFlags)
+				db.dirty = true
+			}
+		} else {
+			db.size += len(l.getValue(hdr.oldValue))
+		}
+
+		l.moveCursor(&cursor, &hdr)
```
Comment (coocood) [inline kv/new_memdb_arena.go:297]: moveBackCursor may be more accurate.

File: kv/new_memdb_iterator.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+func (i *memdbIterator) Next() error {
+	for {
+		if i.reverse {
+			i.curr = i.db.predecessor(i.curr)
+		} else {
+			i.curr = i.db.successor(i.curr)
+		}
+
+		// We need to skip persistent flags only nodes.
+		if i.includeFlags || i.curr.isNull() || !i.curr.vptr.isNull() {
```
Comment (coocood) [inline kv/new_memdb_iterator.go:81]: Extract a `isFlagOnly` method is easier to read.

File: kv/new_memdb_iterator.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	if cmp <= 0 && !y.isNull() {
+		i.curr = i.db.predecessor(y)
+		return
+	}
+	i.curr = y
+}
+
+func (i *memdbIterator) fixPosition() {
```
Comment (coocood) [inline kv/new_memdb_iterator.go:149]: The name does not carry much information, maybe inline this method is easier to understand, like
```
if i.isFlagOnly() && !i.includeFlags {
    err := i.Next()
    _ = err
}

```

**Comment by coocood** [comment]
LGTM

**Comment by bobotu** [comment]
PTAL @tiancaiamao

**Comment by bobotu** [comment]
PTAL @tiancaiamao

**Pull request merged**